cpp/src/parquet/encoding.cc

```diff
@@ -1482,117 +1387,60 @@ class PlainByteArrayDecoder : public PlainDecoder<ByteArrayType>,
         }
-        helper.UnsafeAppend(data_ + sizeof(uint32_t), value_len);
+        helper.UnsafeAppend(data_ + 4, value_len);
         data_ += increment;
         len_ -= increment;
         ++values_decoded;
       } else {
         helper.UnsafeAppendNull();
       }
-      bit_reader.Next();
-    }
+      ++i;
+      return Status::OK();
+    };
+
+    RETURN_NOT_OK(VisitNullBitmapInline(valid_bits, valid_bits_offset, num_values,
+                                        null_count, std::move(decode_value)));
 
     num_values_ -= values_decoded;
     *out_values_decoded = values_decoded;
     return Status::OK();
   }
 
-  Status DecodeArrowDenseNonNull(int num_values,
-                                 typename EncodingTraits<ByteArrayType>::Accumulator* out,
-                                 int* values_decoded) {
-    ArrowBinaryHelper helper(out);
-    num_values = std::min(num_values, num_values_);
-    RETURN_NOT_OK(helper.builder->Reserve(num_values));
-    RETURN_NOT_OK(helper.builder->ReserveData(
-        std::min<int64_t>(len_, helper.chunk_space_remaining)));
-    for (int i = 0; i < num_values; ++i) {
-      // For compiler warnings on unsigned/signed arithmetic.
-      auto int32_s = static_cast<int32_t>(sizeof(int32_t));
-
-      auto value_len = arrow::util::SafeLoadAs<int32_t>(data_);
-      if (ARROW_PREDICT_FALSE(value_len < 0 || value_len > INT32_MAX - int32_s)) {
-        return Status::Invalid("Invalid or corrupted value_len '", value_len, "'");
-      }
-
-      auto increment = int32_s + value_len;
-      if (ARROW_PREDICT_FALSE(len_ < increment)) {
-        ParquetException::EofException();
-      }
-      if (ARROW_PREDICT_FALSE(!helper.CanFit(value_len))) {
-        // This element would exceed the capacity of a chunk
-        RETURN_NOT_OK(helper.PushChunk());
-        RETURN_NOT_OK(helper.builder->Reserve(num_values - i));
-        RETURN_NOT_OK(helper.builder->ReserveData(
-            std::min<int64_t>(len_, helper.chunk_space_remaining)));
-      }
-      helper.UnsafeAppend(data_ + sizeof(uint32_t), value_len);
-      data_ += increment;
-      len_ -= increment;
-    }
-
-    num_values_ -= num_values;
-    *values_decoded = num_values;
-    return Status::OK();
-  }
-
   template <typename BuilderType>
   Status DecodeArrow(int num_values, int null_count, const uint8_t* valid_bits,
                      int64_t valid_bits_offset, BuilderType* builder,
                      int* out_values_decoded) {
     RETURN_NOT_OK(builder->Reserve(num_values));
-    arrow::internal::BitmapReader bit_reader(valid_bits, valid_bits_offset, num_values);
     int values_decoded = 0;
-    for (int i = 0; i < num_values; ++i) {
-      if (bit_reader.IsSet()) {
-        // For compiler warnings on unsigned/signed arithmetic.
-        auto int32_s = static_cast<int32_t>(sizeof(int32_t));
 
+    auto decode_value = [&](bool is_valid) {
+      if (is_valid) {
+        if (ARROW_PREDICT_FALSE(len_ < 4)) {
```

Comment:
Coding style: please reintroduce the constant (named `kPrefixSize` maybe) or use 
```suggestion
        if (ARROW_PREDICT_FALSE(len_ < sizeof(int32_t))) {
```
rather than `4`

Reply:
-1. `sizeof(int32_t)` is pointless pedantry.